Introduce IMessageReceiverFactory

andreminelli opened the pull request:
The current implementation creates all MessageReceivers instances when starting.

I think could be very useful to allow clients to override this behavior. 
For instance, in order to produce instances scoped by message for instance: imagine a context provider which can offer the message, the contact for this message and other properties related to current user.

To allow this scenario I have added IMessageReceiverFactory, which is used to get the receiver before processing each request. If no implementation could be found by the IServiceProvider, a default implementing the current logic will be used.

Some tests are intentionally failing to demonstrate the impacts of this change. If this request manages to be accepted they should probably be deleted.

andrebires commented:
Thanks for your pull request! This is indeed a very useful feature. May I suggest two things before merging it to master?

1. Make the feature be opt-in. It is more safe for us if the behavior only changes if the developer set a flag for it (for instance, it can be a `lifetime` property in the receiver registration on application.json with the default value `singleton` and an optional value `scoped`). In the `singleton`, the instance is created immediately, just like before. And with the `scoped`, using proposed behavior instead.

2. Implement the same behavior in the [new SDK version](https://github.com/takenet/blip-sdk-csharp) first. It will be probably just a copy-and-paste. Our intent is to maintain this repository only for bug fixes from now on (to encourage the usage of the new version), but of course we will accept this pull request.

andreminelli commented:
You are absolutely right! I have just pushed changes with the `lifetime` property.
I will check the new SDK version.

andrebires commented on src/Takenet.MessagingHub.Client/Host/MessageApplicationReceiver.cs:
```diff
@@ -17,5 +17,13 @@ public class MessageApplicationReceiver : ApplicationReceiver
+        /// - singleton (default)
+        /// - scoped (an instance per message request)
+        /// </summary>
+        public string Lifetime { get; set; }
```
Create an enum for these values

andreminelli replied:
Done

andrebires commented on src/Takenet.MessagingHub.Client/Host/Bootstrapper.cs:
```diff
@@ -559,5 +585,64 @@ protected SetStateEnvelopeReceiver(IEnvelopeReceiver<T> receiver, IStateManager
+                }
+                finally
+                {
+                    await _messageReceiverFactory.ReleaseAsync(receiver);
```
ConfigureAwait(false) is missing

andreminelli replied:
Added

andrebires commented on src/Takenet.MessagingHub.Client/Host/Bootstrapper.cs:
```diff
@@ -559,5 +585,64 @@ protected SetStateEnvelopeReceiver(IEnvelopeReceiver<T> receiver, IStateManager
+
+            public MessageReceiverFactory(IServiceProvider provider)
+            {
+                this._provider = provider;
```
`this.` is redundant

andreminelli replied:
Gone.

andrebires commented on src/Takenet.MessagingHub.Client/Host/Bootstrapper.cs:
```diff
@@ -504,6 +503,33 @@ private static void RegisterTunnelReceivers(Application application)
 
+        private static Task<IMessageReceiver> BuildByLifetimeAsync(string lifetime, Type receiverType, IDictionary<string, object> settings, IServiceContainer serviceContainer)
+        {
+            if ("scoped".Equals(lifetime, StringComparison.InvariantCultureIgnoreCase))
```
Introduce an enum instead of a string for the lifetime value

andreminelli replied:
Magic string removed

andreminelli commented:
I have fixed the issues you pointed out, and I have added another property to main `Application`, `DefaultMessageReceiverLifetime`, so we can change all receivers easily.

Please review again. As soon as this version is approved I will apply the same changes to new SDK...

andrebires commented:
Merged, thanks!
